src/mjolnir/landmark_builder.cc

```diff
@@ -220,6 +228,80 @@ void LandmarkDatabase::release_prepared_stmt() {
   LOG_INFO("released prepared statements");
 }
 
+// Task 2: parse landmarks
+// anonymous namespace?
+struct landmark_callback : public OSMPBF::Callback {
+public:
+  landmark_callback(LandmarkDatabase& db) : db_(db) {
+  }
+  virtual ~landmark_callback() {
+  }
+
+  virtual void
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    Landmark landmark;
+
+    for (const auto& tag : tags) {
+      if (tag.first == "amenity") {
+        // if amenity is empty will return LandmarkType::null
+        landmark.type = string_to_landmark_type(tag.second);
+      }
+      if (tag.first == "name" && !tag.second.empty()) {
+        landmark.name = tag.second;
+      }
+      landmark.lng = lng;
+      landmark.lat = lat;
+    }
+    // insert parsed landmark directly into database
+    db_.insert_landmark(landmark);
+  }
+
+  virtual void changeset_callback(const uint64_t changeset_id) override {
+  }
+
+  virtual void way_callback(const uint64_t /*osmid*/,
+                            const OSMPBF::Tags& /*tags*/,
+                            const std::vector<uint64_t>& /*nodes*/) override {
+    LOG_WARN("way callback shouldn't be called!");
+  }
+
+  virtual void relation_callback(const uint64_t /*osmid*/,
+                                 const OSMPBF::Tags& /*tags*/,
+                                 const std::vector<OSMPBF::Member>& /*members*/) override {
+    LOG_WARN("relation callback shouldn't be called!");
+  }
+
+  LandmarkDatabase& db_;
+};
```

Comment:
In admin builder the callback struct is put in anonymous namespace. should I do the same here? don't know the benefits though

Reply:
Jep we should do the same here. In this case there'd be hardly a benefit, other than making it clear that it's not part of the public interface and rather implementation. If it's inside our named namespaces, I'd expect the callback struct's definition to be inside the header file too so I could import it into other files. Since we only use that struct in this .cc file, we should put it into a anonymous namespace.

Reply:
unnamed namespace is best for two reasons:

1. this is not in the valhalla/mjolnir namespace because we want it to remain an implementation detail of the library and not something in the public interface
2. adding it to a namespace makes it local to that translation unit and as such avoids potential naming clashes when linking